### PR TITLE
Don't derive type var constraints from type param bounds

### PR DESCRIPTION
**Pull request by adriaanm**
This overly constrains type checking, as reported in #6739 and scala/bug#10819.

### CODE REVIEW

**Comment by dwijnand** [comment]
Looks like you need to be less ambitious. Or differently ambitious.

```
[error] /home/travis/build/scala/scala/src/library/scala/collection/concurrent/TrieMap.scala:986: type mismatch;
[error]  found   : scala.collection.concurrent.TrieMap[K,V]
[error]  required: scala.collection.concurrent.TrieMap[K,V] with scala.collection.mutable.Growable[Elem]
[error]   def newBuilder[K, V] = new GrowableBuilder(empty[K, V])
[error]                                                   ^
```

(one of a number of failures in Travis CI)

**Comment by adriaanm** [comment]
As usual, my ambitions are being adjusted towards the more ad-hoc. I think I have a compromise brewing.

**Comment by adriaanm** [comment]
Bah, it passes the test suite, though.

**Comment by adriaanm** [comment]
As usual, too much code has started to rely on this. Not sure we can do much here.

**Comment by adriaanm** [comment]
No, we really can't let this escape into the wild. It's only been enabled since M3, so not sure why so many tests fail. Must be doing something else wrong.

**Comment by milessabin** [comment]
@adriaanm that looks a lot nicer.

> TODO: why do we need the additional hi-bound? see sip23-widen

What's the issue with sip23-widen?

**Comment by milessabin** [comment]
@adriaanm Ahh ... OK, I see. We need to propagate the `<: Singleton` from the symbol to the constraint so that it's available [here](https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala#L258). Just stopping widening isn't sufficient.

I'm not saying there isn't a better way to do that ...

**Comment by milessabin** [comment]
Hmm ... or perhaps not. Something must be being propagated from the constraint `hiBounds` to some other type variable somewhere.

**Comment by adriaanm** [comment]
Ok, thanks for checking & clarifying. I think that makes sense then.

File: test/files/run/t10819.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    """.stripMargin
+
+  def show(): Unit = {
+    val baos = new java.io.ByteArrayOutputStream()
+    Console.withOut(baos)(Console.withErr(baos)(compile()))
+    val out = baos.toString("UTF-8")
+
+    val inferredType = out.lines.filter(_.contains("C.this.id[")).map(_.trim).toList
+    assert(inferredType.length == 1)
+    assert(inferredType.forall(_.startsWith("C.this.id[Int]")), inferredType)
```
Comment (lrytz) [inline test/files/run/t10819.scala:21]: could also be a ReplTest, the transcript in the check file would show the inferred type

File: test/files/pos/S1.scala [hunk 2]
```diff
@@ -8,24 +8,6 @@
...
-  def foo[T <: this.type](x: T) = x
-  def f = foo[this.type](this)
+    def foo[T <: this.type](x: T) = x;
+    foo[this.type](this);
 }
-
-class S2() {
-  def foo[T <: this.type](x: T) = x
-  def f = foo(this)
-}
```
Comment (lrytz) [inline test/files/pos/S1.scala]: Why was this removed? It still compiles after this PR.

It would also be a good opportunity to rename the file of this test :-) And the comment in the beginning looks stale.

**Review by lrytz — COMMENTED** [review]
Change LGTM

File: test/files/pos/S1.scala (hunk 2 above)
Comment (adriaanm) [reply test/files/pos/S1.scala]: Hmm, the test failed on my machine, but maybe that was spurious.

Comment (lrytz) [reply test/files/pos/S1.scala]: I tried with `scala-pr 6789` in the repl.

Comment (adriaanm) [reply test/files/pos/S1.scala]: Maybe I had a stray class file or something. Pushed a revert for it
